design/one-pager-trait-workload-interaction-mechanism.md

```diff
@@ -0,0 +1,198 @@
+corresponding `workloadDefinition` CR following an 
+[OAM convention](https://github.com/oam-dev/spec/blob/master/3.workload.md#definitionref). The
+conventions says that the name of the `workloadDefinition` CR is the name of the `workload` CRD
+it refers to. For example, the name of the `workloadDefinition` CR that refers to a
+`containerizedworkloads.core.oam.dev` CRD is exactly `containerizedworkloads.core.oam.dev` as
+well. In our example, the applicationConfig controller will emit the following `workload` CR with all
+ its child resources GVK in its metadata.
+    ```yaml
+    apiVersion: standard.oam.dev/v1alpha2
+    kind: Mydb
+    metadata:
+      name: mydb-example
+      annotations:
+        core.oam.dev/childResources: "{ apps/v1.deployment, core/v1.service}"
+    ```
+
+3. ApplicationConfig controller add annotations containing the workload information to the emitted
+`trait` CR. The application controller actually generates the workload CR so it has all the
+information. In our example, the applicationConfig controller will emit the following `trait` CR with
+the GVK and name of the workload the `ManualScalerTrait` is applied to.
+    ```yaml
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ManualScalerTrait
+    metadata:
+      annotations:
+        core.oam.dev/workloadGroupVersion: standard.oam.dev/v1alpha2
+        core.oam.dev/workloadKind: Mydb
+        core.oam.dev/workloadName: mydb-example
+    ```
+4. OAM runtime will provide a helper library. The library follows the following logic to help a
+ trait developer locate the resources for the trait to modify
+    1. Get the corresponding `workload` instance from the Kubernetes cluster with the information
+     contained in the annotation field inserted by the application controller in the `trait` CR.
+    2. Fetch all the `childResources` values inserted by the application controller in the
+     annotation field in the `workload` instance. 
+    3. List each child resource by its GVK and filter by owner reference. Here, we assume that
+     all the child resources that the workload controller generates have an owner reference field
+      pointing back to the workload instance.
+
+## Impact to the existing system
+Here are the impact of this mechanism to the existing controllers
+- ApplicationConfiguration: This mechanism requires changes in the
+ applicationConfiguration controller.
+- Workload: This mechanism does not affect workload implementation.
+- Trait: This mechanism is optional so all existing trait controller still works. This mechanism
+requires modification for any existing traits implementation that wants to take advantage of
+extensibility of OAM. Any traits that only applies to a certain type of workload don't need to
+use this mechanism.
+- WorkloadDefinition: We can modify the existing workloadDefinition if needed.
+
+## Alternative approach
+1. One alternative approach is that we can make the applicationConfiguration controller watch all
```

Comment:
I found this proposal a little confusing to follow. Specially because it adds complexity to AppConfig. Can we have an alternative approach where the controller automatically annotates the child resources with a reference to the parent resource via an annotations? This way, the generated child can be retrieved via querying the annotations. This way, there is no change in the CRDs and this becomes a spec for the controller only. OAM users are not even aware that these annotations are needed since they are auto-generated.

I might be missing something here but it feels like this proposal is making implementation simpler on the expense of increasing the user's language surface.

Reply:
Ryan and I had a call on this. Here are the clarifying points for this proposal:

1. We want to allow existing operators to be used with little or no change, so adding requirements of automatically generating annotations is against that principle.

2. Only the WorkloadDefinition will be “against” the spec by adding childResources property, the other changes are in the generated resources. So, the OAM resources are virtually unchanged.

3. We have a path forward in case there is a more complex generated resources and labeling can be used - no need to implement that now but we have it as a way out if that becomes an issue.

Given the points above. I am fine with this proposal.

Special thanks to Ryan for driving this change and creating clarity as we review it.